### PR TITLE
Add algorithm resource estimation

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR creates a feature branch for the algorithm resource estimation functionality.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -4,6 +4,90 @@
...
+  error = 0.01
+  qml.resource.FirstQuantization.qubit_cost(n, eta, omega, error)
+  ```
+  
+  ```pycon
+  4377
+  ```
+
+  In addition to the number of non-Clifford gates and logical qubits, some other quantities such as
+  the 1-Norm of the Hamiltonian and double factorization of the second-quantized Hamiltonian can be
```
Comment (ixfoduap) [inline doc/releases/changelog-dev.md:88]:
```suggestion
  the 1-norm of the Hamiltonian and double factorization of the second-quantized Hamiltonian can be
```

**Review by ixfoduap — APPROVED** [review]

**Review by mojishoki — APPROVED** [review]

**Review by obliviateandsurrender — APPROVED** [review]

File: pennylane/resource/first_quantization.py [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+            raise ValueError("The number of electrons must be a positive integer.")
+
+        if omega <= 0:
+            raise ValueError("The unit cell volume must be a positive number.")
+
+        if error <= 0.0:
+            raise ValueError("The target error must be greater than zero.")
+
+        if not isinstance(charge, int):
+            raise ValueError("system charge must be an integer.")
```
Comment (antalszava) [inline pennylane/resource/first_quantization.py:562]: I see a Codecov box here, is this case covered by the tests?

File: pennylane/qchem/factorization.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for two-electron tensor factorization.
+"""
+from pennylane import numpy as np
+
+
+def factorize(two, tol_factor=1.0e-5, tol_eigval=1.0e-5):
```
Comment (antalszava) [inline pennylane/qchem/factorization.py:20]: Minor: looking at the signature, the argument `two` seems a bit too general. Could it be renamed?

**Review by antalszava — COMMENTED** [review]
Very exciting! :partying_face: :1st_place_medal: 

Did a quick skim through, one thing stood out for the coverage.

File: pennylane/resource/first_quantization.py (hunk 2 above)
Comment (soranjh) [reply pennylane/resource/first_quantization.py:562]: This [line](https://github.com/PennyLaneAI/pennylane/pull/2644/files/6294a6d812b8ef6481a8229940525ea5020c7907#diff-d830cf664dc2d8efcb6875b068e210cd016651ba50ea2e4cd37546bd425b489aR198) should cover that condition.

**Review by Jaybsoni — APPROVED** [review]
Looks great @soranjh! Just one clarification, why did we decide to implement `FirstQuantization()` and `DoubleFactorization()` as subclasses of `Operation`? Will they somehow be applied to a state in a quantum circuit context?

**Comment by soranjh** [comment]
> Looks great @soranjh! Just one clarification, why did we decide to implement `FirstQuantization()` and `DoubleFactorization()` as subclasses of `Operation`? Will they somehow be applied to a state in a quantum circuit context?

We want to make them template-like with the possibility of having a device in the future that performs estimation instead of execution on these templates.

**Pull request merged**